### PR TITLE
DAOS-12398 object: Check punch epoch before migration punch

### PR DESCRIPTION
**Pull request by wangdi1**
Since object enumeration might return next punch epoch of
the current object/key/rec, so let's compare the punch
epoch with rebuild/reintegration stable epoch before
migrate punch.

Do not allow reint/drain/extend if there are other rebuild
job going on to avoid screwing up the object layout.

Fix converting task error reset in dc_tx_check_update()

Required-githooks: true
Allow-unstable-test: true
Test-tag: pr rebuild ec
Signed-off-by: Di Wang <di.wang@intel.com>
Co-authored-by: Fan Yong <fan.yong@intel.com>
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by daltonbohning — COMMENTED** [review]
Since this touches core code, we should run more tests than just:
`Test-tag: OSAOnlineReintegration`

For the core changes:
`Test-tag: object osa rebuild`
For the changes to `src/tests/suite/daos_rebuild.c` and `src/tests/suite/daos_rebuild_simple.c`
`Test-tag: daos_test` or `Test-tag: pr`

In total, I'd recommend:
`Test-tag: pr object osa rebuild`

**Comment by wangdi1** [comment]
> Since this touches core code, we should run more tests than just: `Test-tag: OSAOnlineReintegration`
> 
> For the core changes: `Test-tag: object osa rebuild` For the changes to `src/tests/suite/daos_rebuild.c` and `src/tests/suite/daos_rebuild_simple.c` `Test-tag: daos_test` or `Test-tag: pr`
> 
> In total, I'd recommend: `Test-tag: pr object osa rebuild`

Oh, this is for debugging purpose for now, since I can not reproduce the issue locally, so need repeat this specific tests to collect necessary log.

**Comment by daltonbohning** [comment]
> Oh, this is for debugging purpose for now, since I can not reproduce the issue locally, so need repeat this specific tests to collect necessary log.

Ah... gotcha! :)

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
failure due to DAOS-11964

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

File: src/tests/ftest/control/dmg_pool_query_ranks.py [hunk 1]
```diff
@@ -141,8 +142,21 @@ def test_pool_query_ranks_mgmt(self):
...
         for rank in all_ranks:
             self.log.debug("Reintegrating rank %d", rank)
-            self.pool.reintegrate(rank)
 
+            counter = 0
+            cmd_succeed=False
+            while counter < 3:
+                try:
+                    result = self.pool.reintegrate(rank)
+                except CommandFailure:
```
Comment (jolivier23) [inline src/tests/ftest/control/dmg_pool_query_ranks.py:151]: This comes from exception_utils so probably need

from exception_utils import CommandFailure

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**